witness/golang/client/http/witness_client.go

```diff
@@ -64,29 +68,36 @@ func (w Witness) GetLatestCheckpoint(ctx context.Context, logID string) ([]byte,
 }
 
 // Update attempts to clock the witness forward for the given log ID.
-func (w Witness) Update(ctx context.Context, logID string, cp []byte, proof [][]byte) error {
+func (w Witness) Update(ctx context.Context, logID string, cp []byte, proof [][]byte) ([]byte, error) {
 	reqBody, err := json.MarshalIndent(&wit_api.UpdateRequest{
 		Checkpoint: cp,
 		Proof:      proof,
 	}, "", " ")
 	if err != nil {
-		return fmt.Errorf("failed to marshal update request: %v", err)
+		return nil, fmt.Errorf("failed to marshal update request: %v", err)
 	}
 	u, err := w.URL.Parse(fmt.Sprintf(wit_api.HTTPUpdate, logID))
 	if err != nil {
-		return fmt.Errorf("failed to parse URL: %v", err)
+		return nil, fmt.Errorf("failed to parse URL: %v", err)
 	}
-	req, err := http.NewRequest("POST", u.String(), bytes.NewReader(reqBody))
+	req, err := http.NewRequest("PUT", u.String(), bytes.NewReader(reqBody))
 	if err != nil {
-		return fmt.Errorf("failed to create request: %v", err)
+		return nil, fmt.Errorf("failed to create request: %v", err)
 	}
 	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
 	if err != nil {
-		return fmt.Errorf("failed to do http request: %v", err)
+		return nil, fmt.Errorf("failed to do http request: %v", err)
 	}
 	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read body: %v", err)
+	}
 	if resp.StatusCode != 200 {
-		return fmt.Errorf("bad status response: %s", resp.Status)
+		if resp.StatusCode == 409 {
+			return body, ErrCheckpointTooOld
```

Comment:
Yes, the user of this API is a feeder. The code at https://github.com/mhutchinson/trillian-examples/commit/ec8a8cf621146bc568ff5f1a79c56897b015dc76#diff-46156c63664b4ba07bae963a1e6f14292f599a066911172713c247717dea42e7R141 may be useful here. I've implemented this the other way around (i.e. check that there is an error and it isn't CheckpointTooOld, in which case exit. Otherwise, parse the bytes)